ApiGen Plugin

wandersonwhcr opened the pull request:
Olá a todos! :smiley:

# ApiGen Plugin `api_gen`

I develop a super simple ApiGen plugin according to #143. You'll only need to add an `apigen.neon` configuration in the root of your project.

More information in https://github.com/apigen/ApiGen.

## Notices

1. I don't know if my `composer.lock` change is right. Future problems with merging? Probably;
2. Plugin supports only `apigen >= 4.*`. Composer `suggest` option don't support versions;
3. Stage "success" and plugins without options cause PHP Fatal Error. 8ca0a04 and 86157cc. Fixed;
4. I called a `chdir` before and after binary execution;
5. I can't use vfs to create a project structure because `apigen` is called with `proc_open`; and
6. My PR will only pass in PHPCI if PR #749 55a2b79 is applied; thanks @corretge.

## Usage

Add to your `phpci.yml`:

```
success:
    api_gen:
```

wandersonwhcr commented:
PHPMD problem. I'll check soon.

wandersonwhcr commented:
gotcha!

wandersonwhcr commented:
## Notices

6.2 or PR #758

dancryer commented:
@wandersonwhcr This looks great, thanks :) 

I'm intrigued what the changes to the Builder class are for?

wandersonwhcr commented:
@dancryer I created tests about it: `Tests/PHPCI/Build/BuilderTest.php`.

If we configure an empty stage, PHP generate a fatal error, because it is expecting an array and receives a null, exactly here:

https://github.com/Block8/PHPCI/blob/master/PHPCI/Plugin/Util/Executor.php#L57

If we configure an empty `test` and `allow_failures`, PHP automagically configures `$options` variable to array at line 53.

Example:

```yaml
success:
```

or

```yaml
success:
    api_gen:
```

Thinking about it now, I just need to create another test, combining Builder and PluginExecutor. Sorry about that.

dancryer commented:
Hi @wandersonwhcr - Please could you rebase this change on top of the current master branch? :)

Otherwise, :+1: from me.

mavimo commented on Tests/PHPCI/Build/BuilderTest.php:
```diff
@@ -0,0 +1,37 @@
+{
+    public function testExecute_EmptyStageConvertNullToArray()
+    {
+        $builder = $this
```
Move builder in a class variable and configure in ```setUp()```.

wandersonwhcr replied:
Ack. Commit 5c2a4e77.

mavimo commented on Tests/PHPCI/Build/BuilderTest.php:
```diff
@@ -0,0 +1,37 @@
+            ->disableOriginalConstructor()
+            ->getMock();
+
+        $config = (new YamlParser())->parse("stage:\n");
```
This should be only an array?

wandersonwhcr replied:
Verifying the current source code, the `PHPCI\Plugin\Util\Executor::buildPlugin` is not expecting the second parameter as array, because of dd58dd68. I don't know why a plugin configuration must not always be an array, but, I'll remove my changes about it.

mavimo commented on Tests/PHPCI/Build/BuilderTest.php:
```diff
@@ -0,0 +1,37 @@
+            ->disableOriginalConstructor()
+            ->getMock();
+
+        $config = (new YamlParser())->parse("stage:\n    plugin:\n");
```
This should be only an array?

wandersonwhcr commented:
I refactored my code, cleaning some dirty commits and removing my `YamlParser` patch, because dd58dd68 removed the expected array value at `PHPCI\Plugin\Util\Factory::buildPlugin` method.

But now I don't know why my source code don't pass in PHPCI. I cloned your source code and checked out `1.6.0` tag and I installed it. I added a project in PHPCI, pointing to my local path. When I build locally, all works fine.

Can you send me the errors from fd98b20 commit?

mavimo commented:
@wandersonwhcr There is a PHPUnit test that fail:

Failure: PHPCI\Plugin\Tests\ApiGenTest::testSuccess
Failed asserting that false is true.
